### PR TITLE
feat: new command line args & command prompts for Google service file output & write debug symbol scripts to negate the use of "firebase_app_id_file.json"

### PR DESCRIPTION
**Pull request by russellwheatley**
## Description

Allow the option of writing iOS, macOS & android Google service file to a specified directory. How it works:

### Example:
```bash
flutterfire configure  --ios-out=/ios/GoogleService-Info.plist \
  --android-out=/android/google-services.json \ --macos-out=/macos/GoogleService-Info.plist
```

It uses the Flutter project as the root and adds the service file to the path and file name specified in the CLI options. The above example would add the service files to the root of each platform folder:

`cool_flutter_project/ios/GoogleService-Info.plist`
`cool_flutter_project/android/google-services.json`
`cool_flutter_project/macos/GoogleService-Info.plist`

### android

There isn't much to do for android except to write the file. according to the [multi-environment documentation](https://firebase.google.com/docs/projects/multiprojects#support_multiple_environments_in_your_android_application), you simply need to write the service file to the correct directory (e.g. `android/app/development/google-services.json`). 

FFCLI also writes to the `android/build.gradle` & `android/app/build.gradle` which are used across all builds & flavors.

______________________________________________________________________________________________________

### iOS & macOS

- [x] what to do about writing the service file for iOS considering the multiple types of setup.

The approach I'm taking is to prompt the user to pick whether to write the service file for a scheme, target or to simply write to a path without any further configuration.

1. Scheme
Find schemes available and add a build phase script that adds the file to the app bundle. You can have multiple scheme configurations and it will still add to the app bundle. For instance, suppose you were using VGV CLI, and you chose to add a service file to the `development` scheme. It will include the service file the app bundle for the following configurations `Release-development`, `Staging-development` & `Debug-development`.

Example of script added to build phases:

```bash
#!/bin/bash

PLIST_DESTINATION=${BUILT_PRODUCTS_DIR}/${PRODUCT_NAME}.app
GOOGLESERVICE_INFO_PATH=/Users/russellwheatley/projects/devTests/service_testing/ios/some/GoogleService-Info.plist

# Copy GoogleService-Info.plist for appropriate scheme. Each scheme has multiple configurations (i.e. Debug-development, Debug-staging, etc).
# This is why we use *"scheme"*
if [[ "${CONFIGURATION}" == *"development"* ]];
then
    echo "Copying ${GOOGLESERVICE_INFO_PATH} to ${PLIST_DESTINATION}"
    cp "${GOOGLESERVICE_INFO_PATH}" "${PLIST_DESTINATION}"
fi     
```

2. Target
This is more straight forward, a command prompt allows you to select the desired target to include the service file. It will then add it to the "Copy bundle resources" phase for that target.

3. Write with no further configuration
This command prompt will simply add the file to the desired destination, creating necessary directories and is for user who want to do their own config setup.

______________________________________________________________________________________________________
- [x] what to do about `firebase_app_id.json` file write. 

I've removed the need to have a `firebase_app_id.json`. The user now has the option to update their build phases with a script that will upload the debug symbols for Crashlytics for their setup (i.e. target or scheme). If the user selects a project setup that uses schemes for example, the CLI will write a bash script that will run for that specific scheme.

The example below is the script written if the user selects a "development" scheme script:
```bash
#!/bin/bash

# Run upload symbol script for appropriate scheme. Each scheme has multiple configurations (i.e. Debug-development, Debug-staging, etc).
# This is why we use *"scheme"*
if [[ "${CONFIGURATION}" == *"development"* ]];
then
    echo "Running firebase_crashlytics upload debug symbols script for 'development' scheme"
    $PODS_ROOT/FirebaseCrashlytics/upload-symbols --build-phase --validate -ai '1:361461320736:ios:6ecd45b6d0d3126f6f8045'
    $PODS_ROOT/FirebaseCrashlytics/upload-symbols --build-phase -ai '1:361461320736:ios:6ecd45b6d0d3126f6f8045'
fi     
```

The below example is a target script. In this example, the target they selected is "Runner":
```bash
#!/bin/bash

# Run upload symbol script for appropriate target.
if [ "${TARGET_NAME}" == "Runner" ];
then
    echo "Running [firebase_crashlytics] upload debug symbols script for 'Runner' target"
    $PODS_ROOT/FirebaseCrashlytics/upload-symbols --build-phase --validate -ai '1:361461320736:ios:6ecd45b6d0d3126f6f8045'
    $PODS_ROOT/FirebaseCrashlytics/upload-symbols --build-phase -ai '1:361461320736:ios:6ecd45b6d0d3126f6f8045'
fi 
```

Here's an example of the scripts written when using "schemes" setup. They will only be written once for each `flutterfire configure` call:
Screenshot 2022-11-25 at 13 56 31

## Notes
1. macOS "scheme" setup which bundles the 'GoogleService-Info.plist' won't work at the moment because there's a code signing issue. It does work for "target" setup.
2. If you do test this feature, and you choose to add the "crashlytics upload debug symbol" script, you will need to delete the `ios/firebase_app_id_file.json` & `macos/firebase_app_id_file.json` because the script is now added in the build phases.

______________________________________________________________________________________________________

closes https://github.com/invertase/flutterfire_cli/issues/14

## Type of Change

- [X] ✨ `feat` -- New feature (non-breaking change which adds functionality)
- [ ] 🛠️ `fix` -- Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ `!` -- Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 `refactor` -- Code refactor
- [ ] ✅ `ci` -- Build configuration change
- [ ] 📝 `docs` -- Documentation
- [ ] 🗑️ `chore` -- Chore

### CODE REVIEW

File: packages/flutterfire_cli/lib/src/commands/config.dart [hunk 1]
```diff
@@ -126,6 +126,27 @@ class ConfigCommand extends FlutterFireCommand {
       help:
           'Whether to generate the firebase_app_id.json files used by native iOS and Android builds.',
     );
+
+    argParser.addOption(
+      'ios-out',
+      valueHelp: 'pathForIosConfig',
+      help:
+          'Where would you like your `Google-Service-Info.plist` file to be written for iOS platform. Useful for different flavors',
```
Comment (Lyokone) [inline packages/flutterfire_cli/lib/src/commands/config.dart:134]:
```suggestion
          'Where to write the `Google-Service-Info.plist` file for the iOS platform. Useful for different flavors',
```

I would use a shorter sentence, less polite 😅

File: packages/flutterfire_cli/lib/src/commands/config.dart [hunk 2]
```diff
@@ -119,12 +118,32 @@ class ConfigCommand extends FlutterFireCommand {
...
     );
     argParser.addFlag(
-      'app-id-json',
-      defaultsTo: true,
+      'debug-symbols-script',
       hide: true,
-      abbr: 'j',
+      abbr: 'd',
+      help:
+          "Whether you want an upload Crashlytic's debug symbols script adding to the build phases of your iOS project.",
```
Comment (Lyokone) [inline packages/flutterfire_cli/lib/src/commands/config.dart:125]:
```suggestion
          "Whether you want an upload Crashlytic's debug symbols script added to the build phases of your iOS project.",
```

File: packages/flutterfire_cli/lib/src/commands/config.dart [hunk 3]
```diff
@@ -119,12 +118,32 @@ class ConfigCommand extends FlutterFireCommand {
...
+      valueHelp: 'pathForIosConfig',
+      help:
+          'Where to write the `Google-Service-Info.plist` file for the iOS platform. Useful for different flavors',
+    );
+
+    argParser.addOption(
+      'macos-out',
+      valueHelp: 'pathForMacosConfig',
+      help:
+          'Where would you like your `Google-Service-Info.plist` file to be written for macOS platform. Useful for different flavors',
```
Comment (Lyokone) [inline packages/flutterfire_cli/lib/src/commands/config.dart:139]: We should probably use the same sentence as iOS

File: packages/flutterfire_cli/lib/src/commands/config.dart [hunk 4]
```diff
@@ -119,12 +118,32 @@ class ConfigCommand extends FlutterFireCommand {
...
+    );
+
+    argParser.addOption(
+      'android-out',
+      valueHelp: 'pathForAndroidConfig',
       help:
-          'Whether to generate the firebase_app_id.json files used by native iOS and Android builds.',
+          'Where would you like your `google-services.json` file to be written for android platform. Useful for different flavors',
```
Comment (Lyokone) [inline packages/flutterfire_cli/lib/src/commands/config.dart:146]: Same thing

File: packages/flutterfire_cli/lib/src/commands/config.dart [hunk 5]
```diff
@@ -170,8 +189,46 @@ class ConfigCommand extends FlutterFireCommand {
...
+    if (updatedMACOSServiceFilePath != null) {
+      return updatedMACOSServiceFilePath;
+    }
+    return argResults!['macos-out'] as String?;
+  }
+
+  // This allows us to update to the required "GoogleService-Info.plist" file name for macOS target or scheme writes.
+  String? updatedMACOSServiceFilePath;
+
+  String? get fullMACOSServicePath {
```
Comment (Lyokone) [inline packages/flutterfire_cli/lib/src/commands/config.dart:206]: Should we have macOS in lower case? Uppercase makes me think of the MAC address of a computer
```suggestion
  String? get fullmacOSServicePath {
```

File: packages/flutterfire_cli/lib/src/commands/config.dart [hunk 6]
```diff
@@ -170,8 +189,46 @@ class ConfigCommand extends FlutterFireCommand {
...
+    if (updatedIOSServiceFilePath != null) {
+      return updatedIOSServiceFilePath;
+    }
+    return argResults!['ios-out'] as String?;
+  }
+
+  // This allows us to update to the required "GoogleService-Info.plist" file name for iOS target or scheme writes.
+  String? updatedIOSServiceFilePath;
+
+  String? get fullIOSServicePath {
```
Comment (Lyokone) [inline packages/flutterfire_cli/lib/src/commands/config.dart:223]: maybe same here?
```suggestion
  String? get fulliOSServicePath {
```

File: packages/flutterfire_cli/lib/src/commands/config.dart [hunk 7]
```diff
@@ -527,17 +642,185 @@ class ConfigCommand extends FlutterFireCommand {
       final xcodeProjFilePath =
           path.join(flutterApp!.iosDirectory.path, 'Runner.xcodeproj');
 
-      final rubyScript =
-          generateRubyScript(googleServiceInfoFile, xcodeProjFilePath);
-
+      // We need to prompt user whether they want a scheme configured, target configured or to simply write to the path provided
       if (Platform.isMacOS) {
```
Comment (Lyokone) [inline packages/flutterfire_cli/lib/src/commands/config.dart:646]: Maybe we can use the same casing as for Platform.isMacOS?

File: packages/flutterfire_cli/lib/src/commands/config.dart [hunk 8]
```diff
@@ -557,17 +867,185 @@ class ConfigCommand extends FlutterFireCommand {
...
-
-        if (result.exitCode != 0) {
-          throw Exception(result.stderr);
+        // We need to prompt user whether they want a scheme configured, target configured or to simply write to the path provided
+        if (fullMACOSServicePath != null) {
+          final fileName = path.basename(fullMACOSServicePath!);
+          final response = promptSelect(
+            'Would you like your macOS $fileName to be associated with your macOS Scheme or Target (use arrow keys & space to select)?',
+            [
+              'Scheme',
```
Comment (Lyokone) [inline packages/flutterfire_cli/lib/src/commands/config.dart:877]: Should we move part of the functions to another file and separate it into named functions? Getting to thousands of lines is probably going to make the function hard to debug

**Review by Lyokone — APPROVED** [review]

**Comment by creativecreatorormaybenot** [comment]
> There isn't much to do for android except to write the file. according to the [multi-environment documentation](https://firebase.google.com/docs/projects/multiprojects#support_multiple_environments_in_your_android_application), you simply need to write the service file to the correct directory (e.g. android/app/development/google-services.json).

The example should be `android/app/**src/**development/google-services.json`, right @russellwheatley?

**Comment by russellwheatley** [comment]
@creativecreatorormaybenot yes, I've updated the description 👍

**Pull request merged**